tests/transforms/test_optimization/test_commute_through_controls_targets.py

```diff
@@ -0,0 +1,305 @@
+
+        def qfunc():
+            qml.PauliZ(wires=2)
+            qml.S(wires=0)
+            qml.CZ(wires=[0, 2])
+
+            qml.CNOT(wires=[0, 1])
+
+            qml.PhaseShift(0.2, wires=2)
+            qml.T(wires=0)
+            qml.PauliZ(wires=0)
+            qml.CRZ(0.5, wires=[0, 1])
+
+        transformed_qfunc = commute_through_controls_targets(qfunc)
+
+        ops = qml.transforms.make_tape(transformed_qfunc)().operations
+
+        names_expected = ["CZ", "PauliZ", "CNOT", "PhaseShift", "CRZ", "S", "T", "PauliZ"]
+        wires_expected = (
+            [Wires([0, 2]), Wires(2), Wires([0, 1]), Wires(2)] + [Wires([0, 1])] + [Wires(0)] * 3
+        )
+
+        compare_operation_lists(ops, names_expected, wires_expected)
+
+    def test_push_behind_mixed_with_matrix(self):
+        """Test that Z-basis gates after controlled-Z-type gates on controls *and*
+        targets get pushed behind."""
+
+        def qfunc():
+            qml.PauliX(wires=1)
+            qml.S(wires=0)
+            qml.CZ(wires=[0, 1])
+            qml.CNOT(wires=[1, 0])
+            qml.PauliY(wires=1)
+            qml.CRY(0.5, wires=[1, 0])
+            qml.PhaseShift(0.2, wires=0)
+            qml.PauliY(wires=1)
+            qml.T(wires=0)
+            qml.RZ(0.2, wires=0)
+            qml.PauliX(wires=1)
+            qml.CRY(0.2, wires=[1, 0])
+
+        transformed_qfunc = commute_through_controls_targets(qfunc)
+
+        original_ops = qml.transforms.make_tape(qfunc)().operations
+        transformed_ops = qml.transforms.make_tape(transformed_qfunc)().operations
+
+        assert len(original_ops) == len(transformed_ops)
+
+        # Compare matrices
+        matrix_expected = compute_matrix_from_ops_two_qubit(original_ops, wire_order=[0, 1])
+        matrix_obtained = compute_matrix_from_ops_two_qubit(transformed_ops, wire_order=[0, 1])
+        assert check_matrix_equivalence(matrix_expected, matrix_obtained)
+
+
+# Example QNode and device for interface testing
+dev = qml.device("default.qubit", wires=3)
+
+
+def qfunc(theta):
+    qml.PauliX(wires=2)
+    qml.S(wires=0)
+    qml.CNOT(wires=[0, 1])
+    qml.PauliY(wires=1)
+    qml.CRY(theta[0], wires=[2, 1])
+    qml.PhaseShift(theta[1], wires=0)
+    qml.T(wires=0)
+    qml.Toffoli(wires=[0, 1, 2])
+    return qml.expval(qml.PauliZ(0))
+
+
+transformed_qfunc = commute_through_controls_targets(qfunc)
+
+expected_op_list = ["PauliX", "CNOT", "CRY", "PauliY", "Toffoli", "S", "PhaseShift", "T"]
+expected_wires_list = [
+    Wires(2),
+    Wires([0, 1]),
+    Wires([2, 1]),
+    Wires(1),
+    Wires([0, 1, 2]),
+    Wires(0),
+    Wires(0),
+    Wires(0),
+]
+
+
+class TestCommuteThroughControlsTargetsInterfaces:
+    """Test that single-qubit gates can be pushed through controlled gates in all interfaces."""
+
+    def test_commute_through_controls_targets_autograd(self):
+        """Test QNode and gradient in autograd interface."""
+
+        original_qnode = qml.QNode(qfunc, dev)
```

Comment:
might be worth explicitly passing `interface="autograd"` in case the default interface changes in future